client/modules/wj-util/src/index.ts

```diff
@@ -0,0 +1,159 @@
+  target: TR,
+  comparator: (element: T, target: TR) => number | boolean,
+  { min = 0, max = haystack.length - 1, precise = true }: SearchOpts = {}
+) {
+  let index = -1
+  while (min <= max) {
+    index = min + ((max - min) >>> 1)
+    const cmp = comparator(haystack[index], target)
+    if (cmp === true || cmp === 0) return { element: haystack[index], index }
+    if (cmp === false) return null
+    if (cmp < 0) min = index + 1
+    else if (cmp > 0) max = index - 1
+  }
+
+  if (!precise) return { element: null, index }
+
+  return null
+}
+
+/** Checks if an array or object is empty. Will return true for non-objects. */
+export function isEmpty(obj: any) {
+  if (!obj) return true
+  if (obj instanceof Array) return obj.length === 0
+  if (obj.constructor === Object) return Object.keys(obj).length === 0
+  return true
+}
+
+/** Creates a type that is the type of `T` if it had a known property `K`. */
+type Has<K extends string, T> = T extends { [P in K]?: infer R }
+  ? Omit<T, K> & Record<K, R>
+  : never
+
+/** Returns if an object `T` has a key `K`, and only returns true if
+ *  the value of that key isn't undefined. */
+export function has<K extends string, T>(
+  key: K,
+  obj: T
+): obj is T extends Record<any, any> ? Has<K, T> : never {
+  if (!isAnyObject(obj)) return false
+  return key in obj && (obj as any)[key] !== undefined
+}
+
+/** Removes all properties assigned to `undefined` in an object. */
+export function removeUndefined<T>(obj: T) {
+  // this wacky approach is faster as it avoids an iterator
+  const keys = Object.keys(obj) as (keyof T)[]
+  for (let i = 0; i < keys.length; i++) {
+    if (obj[keys[i]] === undefined) delete obj[keys[i]]
+  }
+  return obj as { [K in keyof T]: Exclude<T[K], undefined> }
+}
+
+/** Takes a string and escapes any `RegExp` sensitive characters. */
+export function escapeRegExp(str: string) {
+  return str.replace(/[.*+?^${}()|\[\]\\]/g, "\\$&")
+}
+
+/** Checks if a string has any of the provided sigils.
+ *  (e.g. `hasSigil('!string', '!') -> true`) */
+export function hasSigil<T extends string = string>(
+  str: unknown,
+  sigils: string | string[]
+): str is T {
+  if (typeof str !== "string") return false
+  if (typeof sigils === "string") return str.startsWith(sigils)
+  for (const sigil of sigils) if (str.startsWith(sigil)) return true
+  return false
+}
+
+/** Removes sigils from a string recursively. */
+export function unSigil<T extends string>(str: T, sigils: string | string[]): T {
+  if (typeof str !== "string") throw new TypeError("str must be a string")
+  if (typeof sigils === "string") {
+    return (str.startsWith(sigils) ? str.slice(sigils.length) : str) as T
+  } else {
+    for (const sigil of sigils) {
+      if (str.startsWith(sigil)) {
+        return unSigil(str.slice(sigil.length), sigils) as T
+      }
+    }
+  }
+  return str as T
+}
+
+/** Creates a simple pseudo-random ID, with an optional prefix attached. */
+export function createID(prefix = "") {
```

Comment:
```suggestion
export function createID(prefix = ""): string {
```

Reply:
If you want explicit function returns, that can be an eslint rule. Personally, I don't like that rule, because some return types can be really complex and it forces you to get super verbose or split everything into types.

Reply:
tbf it sounds like I'm just complaining about typed languages there, but it stems from how loose JS is. I don't actually know how many people set up their linter to enforce return types - I certainly don't see it that often. Usually I see it recommended because it speeds up the typechecker a little bit.